css/selectors/invalidation/user-action-pseudo-classes-in-has.html

```diff
@@ -37,47 +38,56 @@
 </div>
 <script>
   const tab_key = '\ue004';
-  const hovermeRect = hoverme.getBoundingClientRect();
-  const focusme1Rect = focusme1.getBoundingClientRect();
 
   promise_test(async () => {
     assert_equals(getComputedStyle(subject1).color, "rgb(0, 0, 0)", "subject1 initially black");
     assert_equals(getComputedStyle(subject2).color, "rgb(0, 0, 0)", "subject3 initially black");
 
     await new test_driver
-      .Actions()
-      .pointerMove(hovermeRect.x + 1, hovermeRect.y + 1, {origin: "viewport"})
-      .send();
+        .Actions()
+        .pointerMove(0, 0, {origin: hoverme})
+        .send();
     assert_equals(getComputedStyle(subject1).color, "rgb(0, 0, 255)",
                   "subject1 should be blue");
     assert_equals(getComputedStyle(subject3).color, "rgb(0, 0, 128)",
                   "subject3 should be navy");
 
     await new test_driver
-      .Actions()
-      .pointerMove(hovermeRect.x + 1, hovermeRect.y + 1, {origin: "viewport"})
-      .pointerDown()
-      .send();
+        .Actions()
+        .pointerMove(0, 0, {origin: unhoverme})
+        .send();
+    assert_equals(getComputedStyle(subject1).color, "rgb(0, 0, 0)",
+                  "subject1 should be back to black");
+    assert_equals(getComputedStyle(subject3).color, "rgb(0, 0, 0)",
+                  "subject3 should be back to black");
+
+
+    await new test_driver
+        .Actions()
+        .pointerMove(0, 0, {origin: hoverme})
+        .pointerDown()
+        .send();
     assert_equals(getComputedStyle(subject1).color, "rgb(135, 206, 235)",
                   "subject1 should be skyblue");
     assert_equals(getComputedStyle(subject3).color, "rgb(173, 216, 230)",
                   "subject3 should be lightblue");
 
     await new test_driver
-      .Actions()
-      .pointerUp()
-      .send();
-    assert_equals(getComputedStyle(subject1).color, "rgb(0, 0, 255)",
-                  "subject1 should be blue again");
-    assert_equals(getComputedStyle(subject3).color, "rgb(0, 0, 128)",
-                  "subject3 should be navy again");
+        .Actions()
+        .pointerUp()
+        .pointerMove(0, 0, {origin: unhoverme})
```

Comment:
Could do:
```suggestion
        .Actions()
        .pointerUp()
        .pointerMove(0, 0, {origin: unhoverme})

    // Independent action sequence that should have the last invalidation to be `:has(:hover:active)` -> `:has(:hover)`
    await new test_driver
        .Actions()
        .pointerMove(0, 0, {origin: hoverme})
        .pointerDown()
        .pointerUp()
        .send();
    // ...
```

Reply:
Now that I understand the motivation for the change, I don't think it's important to keep testing the old thing. I think we should treat `:active` and `:hover` as separate, and both are now testing in both directions (matching to not matching and vice versa)

Reply:
I think testing `:has(:hover:active)` &rarr; `:has(:hover)` is perhaps not needed, but you already wrote the test, so let's have it! (I deleted a previous comment saying we don't need this.)